pages/blog/why-build-a-js-engine.md

```diff
@@ -0,0 +1,115 @@
+---
+title: Why build a JavaScript engine?
+description: Addressing the most important elephant in the room.
+date: 2024-07-12
+authors:
+  - name: Aapo Alasuutari
+    url: https://github.com/aapoalas
+---
+
+So, you want to ask the obvious question: Why would someone choose to build a
+new JavaScript engine? Don't we already have the perfect engine in [INSERT
+ENGINE NAME HERE]?
+
+Let me answer that question!
+
+## But first! History!
+
+The very first JavaScript engine to be built was none other than SpiderMonkey:
+Yes, the engine that is still running in Firefox. Other engines followed, many
+of them have already passed into forgotten history but the big ones that
+remained were V8, SpiderMonkey, JavaScriptCore, and Chakra.
+
+The JavaScript world was then a peaceful place where all these lived happily
+together, and all were equal. That is, as long as you don't think about the
+different API support, the competition, and underhanded tactics and all that.
+But at least on the server side things were peaceful: There was only Node.js.
+That is, until Ryan Dahl [nailed his ten theses on the door](https://www.youtube.com/watch?v=M3BM9TB-8yA) of the JSConf EU
+backstage in 2018, and Deno was born.
+
+Deno was still fairly simple: It still runs the same V8 JavaScript engine under
+the hood. But then Bun came along, and it uses JavaScriptCore (of Safari fame)
+so then there were two. But even before that, actually, we had various flavours
+of Node.js that replaced the V8 engine with Microsoft's Chakra engine or with
+Mozilla's SpiderMonkey engine. Oh and actually, the server-side focused QuickJS
+engine has existed long enough to be abandoned and forgotten, get forked, and
+restart development again.
+
+And now that we're dredging up various names, we should mention LibJS which
+powers the Ladybird browser. It is probably the newest (arguable) success story
+of JS engines, having equivalent or higher
+[test262](https://test262.fyi/#|v8,sm,jsc,libjs) passing figures as Google's V8
+and the other major engines. It is, in essence, a fully complete JavaScript
+engine built in only the last few years. It has also acted as the flagship for
+specification-oriented JavaScript engine design, though it was probably not the
+first one to envision of this idea.
+
+Many other engines exist beside these as well: [Boa](https://boajs.dev/), [Kiesel](https://kiesel.dev/), [GraalJS](https://github.com/oracle/graaljs), [engine262](https://engine262.js.org/),
+and then some less traditional ones like [Hermes](https://github.com/facebook/hermes) and [Porffor](https://porffor.dev/). So, if we have such
```

Comment:
Is this implying that engine262 is a "more traditional" engine? It's a JS engine written in JS, meant to be as close to a one-to-one mapping of the specification as possible, so it can be used to e.g. check that a proposal's spec text actually does what it's intended to do.

Reply:
From an implementation point of view, it's probably fairly traditional :)

Reply:
I changed this to "more experimental".